File: src/main/java/com/silverboxsoft/dynamodbtool/controller/inputdialog/AbsDynamoDbSetInputDialog.java
package com.silverboxsoft.dynamodbtool.controller.inputdialog;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javafx.scene.Node;
import javafx.scene.control.Alert;
import javafx.scene.control.Alert.AlertType;
import javafx.scene.control.CheckBox;
import javafx.scene.control.Label;

public abstract class AbsDynamoDbSetInputDialog<T> extends AbsDynamoDbInputDialog<List<T>> {

	protected static final String VALIDATION_MSG_DUP_VALUE = "Same value exists. please change the value";
	protected Node addValueNode;
	private Map<String, T> addAttributeMap;

	public AbsDynamoDbSetInputDialog(List<T> dynamoDbRecord, String dialogTitle) {
		super(dynamoDbRecord, dialogTitle);
		this.setTitle(this.getTitle().concat(String.format("[%s Set]", getTypeString())));
	}

	@Override
	protected List<Integer> getHeaderWidthList() {
		List<Integer> retList = new ArrayList<>();
		retList.add(FILELD_WIDTH);
		retList.add(DEL_COL_WIDTH);
		return retList;
	};

	@Override
	protected List<Node> getHeaderLabelList() {
		Label valTtilelabel = getContentLabel("VALUE", true);
		Label delTtilelabel = getContentLabel("DEL", true);
		List<Node> retList = new ArrayList<Node>();
		retList.add(valTtilelabel);
		retList.add(delTtilelabel);
		return retList;
	}

	@Override
	protected List<List<Node>> getBodyAttributeNodeList() {
		List<List<Node>> retList = new ArrayList<>();
		for (int recIdx = 0; recIdx < getDynamoDbRecordOrg().size(); recIdx++) {
			T attrVal = getDynamoDbRecordOrg().get(recIdx);
			retList.add(getOneBodyAttributeNodeList(recIdx, attrVal));
		}
		return retList;
	}

	@Override
	protected List<Node> getFooterNodeList() {
		T attrVal = getInitAttribute();
		int recIdx = getNewRecIdx();
		addValueNode = getAttrubuteBox(recIdx, attrVal);

		List<Node> retList = new ArrayList<>();
		retList.add(addValueNode);
		retList.add(getAddButton());

		return retList;
	}

	@Override
	protected int getValueColIndex() {
		return 0;
	};

	@Override
	protected List<T> getEditedDynamoDbRecord() {
		List<T> retList = new ArrayList<T>();

		List<List<Node>> currentBodyNodeList = getCurrentBodyNodeList();
		for (List<Node> wkNodeList : currentBodyNodeList) {
			CheckBox delCheck = (CheckBox) wkNodeList.get(1);
			if (delCheck.isSelected()) {
				continue;
			}
			Node valueNode = wkNodeList.get(0);
			retList.add(getCurrentAttrubuteValue(valueNode));
		}
		return retList;
	}

	@Override
	protected List<T> getEmptyAttr() {
		return new ArrayList<>();
	}

	@Override
	protected void actAddNewAttribute() {
		int recIdx = getNewRecIdx();
		String newIdStr = String.valueOf(recIdx);
		T attrVal = getCurrentAttrubuteValue(addValueNode);
		if (!validationCheck(attrVal)) {
			return;
		}
		addAttributeMap.put(newIdStr, attrVal);
		List<Node> nodelList = getOneBodyAttributeNodeList(recIdx, attrVal);
		addAttributeNodeList(nodelList);
	}

	@Override
	final boolean isFinalValidationOk() {
		Set<T> checkSet = new HashSet<T>();
		List<List<Node>> currentBodyNodeList = getCurrentBodyNodeList();
		for (List<Node> wkNodeList : currentBodyNodeList) {
			CheckBox delCheck = (CheckBox) wkNodeList.get(1);
			if (delCheck.isSelected()) {
				continue;
			}
			Node valueNode = wkNodeList.get(0);
			T wkVal = getCurrentAttrubuteValue(valueNode);
			if (wkVal == null) {
				Alert alert = new Alert(AlertType.ERROR, VALIDATION_MSG_INVALID_VALUE);
				alert.showAndWait();
				valueNode.requestFocus();
				return false;
			}
			if (checkSet.contains(wkVal)) {
				Alert alert = new Alert(AlertType.ERROR, VALIDATION_MSG_DUP_VALUE);
				alert.showAndWait();
				valueNode.requestFocus();
				return false;
			}
			checkSet.add(wkVal);
		}
		return true;
	}

	/*
	 * 
	 */
	abstract String getTypeString();

	abstract Node getAttrubuteBox(int recIndex, T attr);

	abstract T getCurrentAttrubuteValue(Node valueNode);

	abstract T getInitAttribute();

	/*
	 * for validation
	 */
	abstract boolean isSameValue(T valA, T valB);

	/*
	 * accessor
	 */
	protected Node getAddValueNode() {
		return addValueNode;
	}

	/*
	 * 
	 */
	protected int getNewRecIdx() {
		return getDynamoDbRecordOrg().size() + getAddAttributeMap().size();
	}

	protected Map<String, T> getAddAttributeMap() {
		if (addAttributeMap == null) {
			addAttributeMap = new HashMap<>();
		}
		return addAttributeMap;
	}

	protected boolean validationCheck(T attrVal) {
		List<List<Node>> currentBodyNodeList = getCurrentBodyNodeList();
		for (List<Node> wkNodeList : currentBodyNodeList) {
			Node valuebox = wkNodeList.get(0);
			T wkAttr = getCurrentAttrubuteValue(valuebox);
			if (isSameValue(wkAttr, attrVal)) {
				Alert alert = new Alert(AlertType.ERROR, VALIDATION_MSG_DUP_VALUE);
				alert.showAndWait();
				addValueNode.requestFocus();
				return false;
			}
		}
		return true;
	}

	private List<Node> getOneBodyAttributeNodeList(int recIdx, T attrVal) {
		Node valueNode = getAttrubuteBox(recIdx, attrVal);
		CheckBox delCheck = new CheckBox();
		delCheck.setId(DEL_ID_PREFIX + String.valueOf(recIdx));

		List<Node> nodeList = new ArrayList<>();
		nodeList.add(valueNode);
		nodeList.add(delCheck);

		return nodeList;
	}
}

File: src/main/java/com/silverboxsoft/dynamodbtool/controller/inputdialog/AbsDynamoDbSetInputDialog.kt
package com.silverboxsoft.dynamodbtool.controller.inputdialog

import javafx.scene.control.CheckBox
import java.util.HashMap
import javafx.scene.control.Alert
import javafx.scene.control.Alert.AlertType
import java.util.HashSet
import javafx.scene.Node
import javafx.scene.control.TextField
import java.util.ArrayList

abstract class AbsDynamoDbSetInputDialog<T>(dynamoDbRecord: List<T>, dialogTitle: String)
    : AbsDynamoDbInputDialog<List<T>>(dynamoDbRecord, dialogTitle) {
    /*
	 * accessor
	 */
    protected var addValueNode: Node = TextField() //dummy
    private var addAttributeMap: MutableMap<String, T> = HashMap()
    override val headerWidthList: List<Int>
        get() {
            val retList: MutableList<Int> = ArrayList()
            retList.add(AbsDynamoDbInputDialog.Companion.FILELD_WIDTH)
            retList.add(AbsDynamoDbInputDialog.Companion.DEL_COL_WIDTH)
            return retList
        }
    override val headerLabelList: List<Node>
        get() {
            val valTitleLabel = getContentLabel("VALUE", true)
            val delTitleLabel = getContentLabel("DEL", true)
            val retList: MutableList<Node> = ArrayList()
            retList.add(valTitleLabel)
            retList.add(delTitleLabel)
            return retList
        }
    override val bodyAttributeNodeList: List<List<Node>>
        get() {
            val retList: MutableList<List<Node>> = ArrayList()
            for (recIdx in dynamoDbRecordOrg.indices) {
                val attrVal = dynamoDbRecordOrg[recIdx]
                retList.add(getOneBodyAttributeNodeList(recIdx, attrVal))
            }
            return retList
        }
    override val valueColIndex: Int
        get() = 0
    override val editedDynamoDbRecord: List<T>
        get() {
            val retList: MutableList<T> = ArrayList()
            val currentBodyNodeList = currentBodyNodeList
            for (wkNodeList in currentBodyNodeList) {
                val delCheck = wkNodeList[1] as CheckBox
                if (delCheck.isSelected) {
                    continue
                }
                val valueNode = wkNodeList[0]
                retList.add(getCurrentAttributeValue(valueNode))
            }
            return retList
        }
    override val emptyAttr: List<T>
        get() = ArrayList<T>()

    override fun getFooterNodeList(): List<Node> {
        addValueNode = getAttributeBox(newRecIdx, initialAttribute)
        val retList: MutableList<Node> = ArrayList()
        retList.add(addValueNode)
        retList.add(addButton)
        return retList
    }

    override fun isValueChanged(): Boolean {
        val edited =  editedDynamoDbRecord
        if (edited.size != dynamoDbRecordOrg.size) {
            return true
        }
        for (recIdx in dynamoDbRecordOrg.indices) {
            val orgVal = dynamoDbRecordOrg[recIdx]
            val curVal = edited[recIdx]
            if (!isSameValue(orgVal, curVal)) {
                return true
            }
        }
        return false
    }
    override fun isAddValueRemain(): Boolean {
        val curVal = getCurrentAttributeValue(addValueNode)
        return !isSameValue(initialAttribute, curVal)
    }

    override fun actAddNewAttribute() {
        val recIdx = newRecIdx
        val newIdStr = recIdx.toString()
        val attrVal = getCurrentAttributeValue(addValueNode)
        if (!validationCheck(attrVal)) {
            return
        }
        addAttributeMap[newIdStr] = attrVal
        val nodeList = getOneBodyAttributeNodeList(recIdx, attrVal)
        addAttributeNodeList(nodeList)
    }

    override val isFinalValidationOk: Boolean
        get() {
            val checkSet: MutableSet<T> = HashSet()
            val currentBodyNodeList = currentBodyNodeList
            var itemCount = 0
            for (wkNodeList in currentBodyNodeList) {
                val delCheck = wkNodeList[1] as CheckBox
                if (delCheck.isSelected) {
                    continue
                }
                val valueNode = wkNodeList[0]
                val wkVal: T = getCurrentAttributeValue(valueNode)
                if (wkVal == null) {
                    val alert = Alert(AlertType.ERROR, AbsDynamoDbInputDialog.Companion.VALIDATION_MSG_INVALID_VALUE)
                    alert.showAndWait()
                    valueNode.requestFocus()
                    return false
                }
                if (checkSet.contains(wkVal)) {
                    val alert = Alert(AlertType.ERROR, VALIDATION_MSG_DUP_VALUE)
                    alert.showAndWait()
                    valueNode.requestFocus()
                    return false
                }
                itemCount += 1
                checkSet.add(wkVal)
            }
            if (itemCount == 0) {
                val alert = Alert(AlertType.ERROR, VALIDATION_MSG_EMPTY)
                alert.showAndWait()
                return false
            }
            return true
        }

    /*
	 * 
	 */
    abstract val typeString: String
    abstract fun getAttributeBox(recIndex: Int, attr: T): Node
    abstract fun getCurrentAttributeValue(valueNode: Node): T
    abstract val initialAttribute: T

    /*
	 * for validation
	 */
    abstract fun isSameValue(valA: T, valB: T): Boolean

    /*
	 * 
	 */
    private val newRecIdx: Int
        get() = dynamoDbRecordOrg.size + getAddAttributeMap().size

    private fun getAddAttributeMap(): Map<String, T> {
        return addAttributeMap
    }

    private fun validationCheck(attrVal: T): Boolean {
        val currentBodyNodeList = currentBodyNodeList
        for (wkNodeList in currentBodyNodeList) {
            val valueBox = wkNodeList!![0]
            val wkAttr = getCurrentAttributeValue(valueBox)
            if (isSameValue(wkAttr, attrVal)) {
                val alert = Alert(AlertType.ERROR, VALIDATION_MSG_DUP_VALUE)
                alert.showAndWait()
                addValueNode.requestFocus()
                return false
            }
        }
        return true
    }

    private fun getOneBodyAttributeNodeList(recIdx: Int, attrVal: T): List<Node> {
        val valueNode = getAttributeBox(recIdx, attrVal)
        val delCheck = CheckBox()
        delCheck.id = AbsDynamoDbInputDialog.Companion.DEL_ID_PREFIX + recIdx.toString()
        val nodeList: MutableList<Node> = ArrayList()
        nodeList.add(valueNode)
        nodeList.add(delCheck)
        return nodeList
    }

    companion object {
        protected const val VALIDATION_MSG_DUP_VALUE = "Same value exists. please change the value"
        protected const val VALIDATION_MSG_EMPTY = "Empty set not allowed. please set one or more values"
    }

    init {
        this.title = this.title + String.format("[%s Set]", typeString)
    }
}